### DISCUSSION TITLE
How to build a model which involves several complicated situations by using command "implies"or any commands

### DISCUSSION THREAD

**Discussion by lJiasong** [yalmip/YALMIP#1325] [Optimization modelling] [answered]
Dear Johan Löfberg
I want to classify and discuss the direction of the lines around the node (currently only consider three lines around a node, that is, the degree of the node is 3), if any two of the lines flow to 1 (1 is used to indicate the flow into the node), the other one The flow direction of the line is -1 (-1 to indicate the outflow node), then in this case, I want to use formula A to model, as you can see, according to the permutation and combination, there will be three situations of this type (1 1 -1 , -1 1 1, 1 -1 1), the remaining cases are modeled using formula B (I think the remaining cases are: the sum of the three line direction values ​​is not equal to 1 to represent).
When I was building a case-by-case discussion model, I encountered some problems, mainly as follows:
First of all, I am not sure whether my logical thinking is consistent with the program I wrote. My binary variable t only appears once in the program, which is used to constrain the flow of the three lines around the node d(i) and only one situation will appear at the same time. , but I always feel that it does not constrain the use of the formula A&B in the subsequent program.
Second, it seems that non-linear statements are not allowed in the order implies? Because a warning pops up at runtime:
![5205cb7bfe157670aceb1fc373c46e8](https://github.com/yalmip/YALMIP/assets/89115142/eeae8a31-dfdd-4e53-8cae-f09eba134bca)
Finally, I attach my code. I sincerely hope that you can give me some suggestions to solve this problem after reading it. I look forward to your reply. I wish you success in scientific research and good health！
PS:The influence of other components on the circuit is not considered in the code, because some components will make it impossible for me to use the formula, but this is not a problem that needs to be considered now. The problem of using the formula caused by the flow direction needs to be solved urgently.
![0f3847a9137284105f7dc5e3e8b9b99](https://github.com/yalmip/YALMIP/assets/89115142/a3b69d84-6b56-4876-ad8f-10ea28ce1119)

**Comment by johanlofberg** [comment]
It is unclear which implies operator is causing your crash as you've supplied a complex piece of code and not simplified for support

But generally, you should always strive to code as explicit disjunction implies(binary1,something)+implies(binary2,something)...+[binary1+binary2+...==1]

Pattern implies(something,something else) or implies(something, binary) is supported but advised against

Pattern implies(something nonlinear,whatever) is probably not supported at all

**Reply by lJiasong** [reply]
Thank for your answer! But I still have some problems. Can pattern implies(something,something nonlinear)  or implies(binary, something nonlinear) be supported? From another perspective,  if the "something" you mentioned is [a==1,b==1,c==1], can it's logical value ( 0 or 1 ) be equivalent to a binary variable to use? That is, whether implies([a ==1,b==1,c==1],whatever) can be supported. I am worried that my programming direction is wrong at the beginning, and the subsequent work will be delayed. I sincerely hope you can give me a few more examples in detail to help me understand the use of this area. My school does not know much about this. Whaterver classmates or teachers, I also hope that I can learn and share it. Thank you again and look forward to your reply.

**Reply by johanlofberg** [reply]
The logical model YALMIP will have to derive just gets very messy when it has to derive it on complex expressions without taking care of context which might make it much more compact

implies([a==1,b==1,c==1], whatever)

I would model as 

implies(case1, [a+b+c ==3 , whatever])+implies(case2, [a+b+c <=2 , whateverthenifanything]) + [case1+case2==1]

**Reply by lJiasong** [reply]
Wow, I think I understand something. Finally, can I understand the "whatever" includes non-linear things even though you didn't advise to do so.

**Reply by johanlofberg** [reply]
yes, but you get a nonlinear integer program, so whether you actually any solver which has any chance at at solving that is another issues (most likely not)

**Comment by lJiasong** [comment]
There is another question, if case1 == 1, then a+b+c must be equal to 3, then “whatever” will be true, right? Similarly, if case2 is true, a+b+c<=2 must also be true, and then The expression is also true (although it will not be true, because what I want to get is a+b+c==3, which leads to case2==0 must be true), am I right to understand this？

**Reply by lJiasong** [reply]
Oh, I used the gurobi solver to solve the model before. At that time, I did not use the implies command, because the previous program could only solve one calculation example. I hope the program can identify some places by itself to improve its versatility, so I added implies Went in but ran into the problem described above. all in all, I will seriously consider your opinions, thank you professor~

**Reply by johanlofberg** [reply, marked as answer]
You can never have a+b+c==3 and case2 true at the same time. That is the whole idea with the disjunction,. Only one out of the two things can happen, and exactly one will have to hold

[a+b+c ==3 , whatever]

[a+b+c <=2 , whateverthenifanything]

**Reply by lJiasong** [reply]
I agree with your point of view, will the later whatever affect the discrimination of the previous expression? For example, “[a+b+c ==3 , whatever]” is established, will the previous ones like "case1" be established because of its establishment? I guess not necessarily, implies is not equivalent to the iff command

**Reply by johanlofberg** [reply]
It is unclear what you are asking. No problems writing iff using implies but in a way that ensures it works, which is not guaranteed when the operators don't have the whole context as you do when you set it up using disjunctions

x<=3 if and only y<=3 should be written as the guaranteed disjunction
implies(a, [x>=3, y>=3]) + implies(b,[x<=3,y<=3]) + [a+b==1]

If y<3 then obviously a cannot be true hence b is true hence x<=3 etc

**Reply by lJiasong** [reply]
OK, I understand. My statement is a little unclear, which caused you trouble. Thank you for your reply. According to your guidance, I will continue to try to modify the program
